implement splits editor - mvp

Eein opened the pull request:
Implements a rough splits editor;
Planning  a refactor soon, but wanted to get some eyes on things.

*Note to onlookers: Please use a backup splits file if you plan on toying with this*

# TODO
- [x] Real Time/Game Time toggle
- [x] Goal Comparison
- [x] Custom Comparisons
- [x] Move to UniqueId for referencing Fields
- [x] Dedicated Save Button on bottom of Details Edit
- [x] Save Button on Edit Split
- [x] `take_user_data` on save and quit

# Wont implement
- File-Based Comparison Import
- Additional Data menu (metadata)

# Notable jank

- `add_splits` function does some strange padding jank to satisfy not having nestable `LinearLayouts` due to a cache bug 
- cleaning up/refactoring modules is pretty much mandatory at this point, but this is a working MVP to test
- theres some wonky error handling but it seems to work for most cases by consuming livesplit errors
- i do alot of assignments for readability, if thats not desired let me konw
- theres some performance issues still, need to dive deeper but its most likely just because i'm using `LinearLayout` for managing alignment 
- minimum 80 chars required to display entire ui (maybe a bit less)
- reused the yucky `find_name`, since it seems like the best way to handle layout rehydration without pulling away management of the state from `livesplit-core`

# Things noticed

- `let globals: GlobalState = s.take_user_data().unwrap();` rips out the current state so it cannot be used further in the loop if necessary it may be better to just use ` let globals = s.user_data::<GlobalState>().unwrap();` everywhere for sanity, unless the idea is to save and quit on save
- future Livesplit core (0.13) requires two minor changes to be up to date

Eein committed:
implement splits editor - mvp

DarkRTA commented on src/lib.rs:
```diff
@@ -8,6 +8,7 @@ use std::sync::atomic::AtomicU64;
 
 pub const CURSIVE_THEME: &str = include_str!("cursive-theme.toml");
 pub const DEFAULT_LAYOUT: &[u8; 6413] = include_bytes!("default_layout.ls1l");
+pub const DEFAULT_SPLITS: &[u8] = include_bytes!("default_splits.lss");
```
please use `Run::default()` instead of doing this

i only have a default layout included as i wanted something custom, and it was far simpler to do it like this instead of building it in code

DarkRTA commented:
> cleaning up/refactoring modules is pretty much mandatory at this point, but this is a working MVP to test

the whole thing is likely to be thrown out and rewritten when we get an actual good terminal library. if you want to do further work in this regard though, i won't mind

> theres some wonky error handling but it seems to work for most cases by consuming livesplit errors

cursive doesn't exactly make this easy. this library is actually garbage

> i do alot of assignments for readability, if thats not desired let me konw

i actually prefer this as opposed to longer lines and deeper nesting

> theres some performance issues still, need to dive deeper but its most likely just because i'm using LinearLayout for managing alignment

was this in a release build? debug builds are slows as hell

> minimum 80 chars required to display entire ui (maybe a bit less)
this is fine

> reused the yucky find_name, since it seems like the best way to handle layout rehydration without pulling away management of the state from livesplit-core

again this library is poorly designed so i'm not gonna be upset over this

> let globals: GlobalState = s.take_user_data().unwrap(); rips out the current state so it cannot be used further in the loop if necessary it may be better to just use  let globals = s.user_data::<GlobalState>().unwrap(); everywhere for sanity, unless the idea is to save and quit on save

current idea is to quit on save as you specify the file on the command line

>  future Livesplit core (0.13) requires two minor changes to be up to date
this is not an issue

DarkRTA commented on src/bin/chronos-splits-editor/main.rs:
```diff
@@ -0,0 +1,646 @@
+    Ok(())
+}
+
+#[rustfmt::skip]
```
please avoid suppressing rustfmt everywhere

Eein replied:
This is mostly because of PaddedView going vertical with its params - I'll look at a cleaner solution (ex: multiple assignment)

DarkRTA commented on src/bin/chronos-splits-editor/main.rs:
```diff
@@ -0,0 +1,646 @@
+        .with_name("details_panel");
+
+    let splits_title = PaddedView::lrtb(0, 0, 0, 1,
+        LinearLayout::horizontal()
```
minor style nit: i would have split this into multiple assignments for readability, but as this is how the examples encourage you to do it, its up to you what you do here. feel free to ignore this one

Eein replied:
I'll refactor it - it will help remove the skip formats

DarkRTA commented:
i am not done reviewing this yet. there is likely more i need to look at

DarkRTA commented:
on the edit splits menu, i would have a dedicated save button at the bottom. it is extremely unclear how to save what you enter otherwise

DarkRTA commented:
also my idea of mvp is to have most features available to edit

there is no way to add comparisons or goal times right now, and probably could be added

Eein commented:
> > theres some performance issues still, need to dive deeper but its most likely just because i'm using LinearLayout for managing alignment
> 
> was this in a release build? debug builds are slows as hell

it was like 2 fps in debug builds - now with the changes its still slow, but not .5fps

Eein commented:
Just a note for myself - need to add a real and game time toggle.

Eein committed:
Refactor to Modules and cleanup

- Refactors to Modules
- Removes default_splits file and uses Run::default()
- General code cleanup

Eein committed:
format and refactor details/layouts

refactor layouts and splits

Eein commented:
> also my idea of mvp is to have most features available to edit
> 
> there is no way to add comparisons or goal times right now, and those likely should be added if its feasible

How are comparisons generally used?
Imported from file or manually entered? 

For file-based comparisons imports, I'd like to avoid making a file browser right for this mvp and tackle it either later or in whatever system gets chosen in the future.

For manual additions, i'll just need to load in custom comparisons at the end and implement right scrolling or a larger size (if thats possible)

I can also revisit the rendering of splits when i get to these.

DarkRTA commented:
I am not too concerned about file import right now. Its a very niche feature that only really works under certain situations.

Eein committed:
adds basic comparisons

Eein committed:
TimingMethods and add Comparison

format

Eein committed:
remove comparison

format

Eein committed:
adds renaming comparisons

Eein committed:
finish adding all comparisons

Eein committed:
move some ids to uniqueid

Eein commented:
notes:
- had to move to full width for comparisons
- use will have to resize to view additional comparisons; theres no reliable way that i've found to keep the title and a horizontal scroll in line

Eein committed:
add cancel buttons and make naming more consistent

Eein committed:
change split editing to use save button

Eein committed:
wrap actions in a panel

Eein committed:
use uniqueid on timing methods

DarkRTA commented:
lgtm with some minor nitpicks

i'm gonna merge this one manually so i can fix them as they are trivial

DarkRTA commented:
Most of it looks good although as of right now its impossible to change the split time.
You might need to rework that menu entirely.

Eein commented:
> Most of it looks good although as of right now its impossible to change the split time.
> You might need to rework that menu entirely.

could have sworn it worked. i'll have a look

DarkRTA commented:
iirc livesplit_core will automatically update some of the times in order to fit certain constraints

if you attempt to update all of them at once, the thing you update last is going to get priority

Eein commented:
i'll be out of town for work for a week - may have time to adjust this during that time, but maybe not - just a heads up

Eein committed:
reimplement split editor

DarkRTA commented:
It is almost there. Only thing that's left is having the times properly refresh in the edit split menu, and then I'll merge it.
Right now the other times do not update correctly at all there, especially if one action results in a lot of other times changing at the same time.

Eein committed:
update split times on edit split window after change

Eein commented:
A little bit rudimentary but it works. 

Let me know if there's anything else you'd like for me to change.

DarkRTA commented:
i'll do some minor fixes on my own and merge this manually tomorrow

CryZe commented:
Yeah a goal is just a normal custom comparison

Eein committed:
adds comparisons to edit split view

Eein commented:
I believe this should work as intended :) :tada:

DarkRTA commented:
i'll get around to reviewing this on the weekend

Eein commented:
Was testing today and gametime/realtime doesn't save.

DarkRTA commented:
i completely forgot about this and im probably just gonna go ahead and merge it assuming it doesn't break anything else
